File: CS/EditorsDemo.Wpf/ModuleResources/CardWinExplorerDataSource.cs
using DevExpress.Xpf.Core;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Windows.Input;
using DevExpress.Utils.Helpers;

namespace EditorsDemo.ModuleResources {
    public class CardWinExplorerDataSource : INotifyPropertyChanged {

        const string Root = "Root";

        SizeIcon sizeType = SizeIcon.Medium;
        string path;
        string searchText;
        readonly Stack<string> UndoStack = new Stack<string>();
        readonly Stack<string> NextStack = new Stack<string>();
        readonly ObservableCollectionCore<CardWinExplorerFileSource> source = new ObservableCollectionCore<CardWinExplorerFileSource>();

        public event PropertyChangedEventHandler PropertyChanged;

        public enum SizeIcon {
            ExtraLarge,
            Large,
            Medium
        }

        public SizeIcon SizeType {
            get { return sizeType; }
            set {
                if (value != sizeType) {
                    sizeType = value;
                    Resize();
                    if (PropertyChanged != null) {
                        PropertyChanged(this, new PropertyChangedEventArgs("Size"));
                        PropertyChanged(this, new PropertyChangedEventArgs("SizeType"));
                    }
                }
            }
        }

        public int Size {
            get {
                switch (SizeType) {
                    case SizeIcon.ExtraLarge:
                        return 256;
                    case SizeIcon.Large:
                        return 128;
                    case SizeIcon.Medium:
                        return 65;
                    default:
                        return 128;
                }
            }
        }

        public string Path {
            get { return path; }
            set {
                string tempPath = value;
                bool fileExist = File.Exists(value);
                if (fileExist) 
                    tempPath = System.IO.Path.GetDirectoryName(tempPath);
                bool directoryExist = Directory.Exists(tempPath);

                if (value != Root && (!directoryExist && !fileExist)) {
                    OpenWeb(value);
                    if (PropertyChanged != null)
                        PropertyChanged(this, new PropertyChangedEventArgs("Path"));
                    return;
                }
                if (path != null)
                    UndoStack.Push(path);
                path = tempPath;
                if(fileExist)
                    OpenFile(value);
                else if(directoryExist || path == Root)
                    OpenFolder(tempPath);
                if (PropertyChanged != null)
                    PropertyChanged(this, new PropertyChangedEventArgs("Path"));
            }
        }
        void OpenFile(string path) {
            try {
                Process.Start(path);
            }
            catch {
            }
        }
        void OpenWeb(string path) {
            try {
                Process.Start("http://" +  path + ".com");
            }
            catch {
            }
        }
        public string SearchText {
            get { return searchText; }
            set {
                searchText = value;
                if (PropertyChanged != null)
                    PropertyChanged(this, new PropertyChangedEventArgs("SearchText"));
            }
        }

        public ObservableCollectionCore<CardWinExplorerFileSource> Source { get { return source; } }

        public CardWinExplorerFileSource CurrentItem { get; set; }

        public ICommand UndoCommand {
            get { return new DevExpress.Mvvm.DelegateCommand(CallUndo, () => UndoStack.Count > 0); }
        }

        public ICommand NextCommand {
            get { return new DevExpress.Mvvm.DelegateCommand(CallNext, () => NextStack.Count > 0); }
        }

        public ICommand OpenCommand {
            get { return new DevExpress.Mvvm.DelegateCommand(() => Open(CurrentItem), () => CurrentItem != null); }
        }

        public ICommand UpCommand {
            get { return new DevExpress.Mvvm.DelegateCommand(CallUp, () => Path != Root); }
        }

        public ICommand PropertyCommand {
            get { return new DevExpress.Mvvm.DelegateCommand(CallExecute, () => CurrentItem != null); }
        }

        public CardWinExplorerDataSource() {
            OpenRoot();
        }

        public void Open(CardWinExplorerFileSource element) {
            if (element.Type == CardWinExplorerFileSource.TypeElement.File)
                FileSystemHelper.ShellExecuteFileInfo(CurrentItem.FullPath, ShellExecuteInfoFileType.Open);
            else
                Path = element.FullPath;
        }

        void OpenFolder(string path, bool clearNextStack = true) {
            Source.Clear();
            Source.BeginUpdate();
            try {
                Mouse.OverrideCursor = Cursors.Wait;
                if (path == Root)
                    OpenRoot();
                else {
                    SizeIcon sizeType = SizeType;
                    int size = Size;
                    foreach (var item in FileSystemHelper.GetSubDirs(path))
                        Source.Add(new CardWinExplorerFileSource(System.IO.Path.Combine(path, item), CardWinExplorerFileSource.TypeElement.Folder, sizeType, size));
                    foreach (var item in Directory.EnumerateFiles(path))
                        Source.Add(new CardWinExplorerFileSource(item, CardWinExplorerFileSource.TypeElement.File, sizeType, size));
                }
                if (clearNextStack)
                    NextStack.Clear();
            }
            catch (UnauthorizedAccessException ex) {
                System.Windows.MessageBox.Show(ex.Message, "Error", System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
                CallUndo();
            }
            finally {
                Mouse.OverrideCursor = null;
                Source.EndUpdate();
            }
        }

        void Resize() {
            try {
                Mouse.OverrideCursor = Cursors.Wait;
                CardWinExplorerFileSource.ClearCache();
                foreach (CardWinExplorerFileSource item in Source)
                    item.Resize(SizeType, Size);
            }
            finally {
                Mouse.OverrideCursor = null;
            }
        }

        void OpenRoot() {
            Source.Clear();
            foreach (var drive in DriveInfo.GetDrives().Where(x => x.DriveType == DriveType.Fixed))
                Source.Add(new CardWinExplorerFileSource(drive.RootDirectory.Name, CardWinExplorerFileSource.TypeElement.Drive, SizeType, Size));
            path = Root;
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs("Path"));
        }

        void CallUp() {
            if (this.Path.Length != 3)
                Path = System.IO.Directory.GetParent(Path).FullName;
            else
                Path = Root;
        }
        void CallExecute() {
            FileSystemHelper.ShellExecuteFileInfo(CurrentItem.FullPath, ShellExecuteInfoFileType.Properties);
        }

        void CallUndo() {
            NextStack.Push(Path);
            string tmp = UndoStack.Pop();
            path = tmp;
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs("Path"));
            OpenFolder(tmp, false);
        }

        void CallNext() {
            string tmp = NextStack.Pop();
            UndoStack.Push(path);
            path = tmp;
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs("Path"));
            OpenFolder(tmp, false);
        }
        public IEnumerable<CardWinExplorerFileSource> Find(string path) {
            if(path == Root) 
                return DriveInfo.GetDrives().Where(x => x.DriveType == DriveType.Fixed).Select(drive => new CardWinExplorerFileSource(drive.RootDirectory.Name, CardWinExplorerFileSource.TypeElement.Drive, SizeType, Size, false));            
            if (string.IsNullOrEmpty(path) || !IsPathValid(path))
                return Enumerable.Empty<CardWinExplorerFileSource>();           
            if(System.IO.Directory.Exists(path))
                path += System.IO.Path.DirectorySeparatorChar;
            var directory = System.IO.Path.GetDirectoryName(path) ;
            if (string.IsNullOrEmpty(directory))
                directory = System.IO.Path.GetPathRoot(path);
            string filter = System.IO.Path.GetFileName(path);

            var directories = FileSystemHelper.GetSubDirs(directory)
                .Where(x => x.StartsWith(filter, true, CultureInfo.InvariantCulture))
                .Select(x => new CardWinExplorerFileSource(System.IO.Path.Combine(directory, x), CardWinExplorerFileSource.TypeElement.Folder, SizeType, Size, false));
            var files = Directory.EnumerateFiles(directory, filter + "*")
                .Select(x => new CardWinExplorerFileSource(x, CardWinExplorerFileSource.TypeElement.File, SizeType, Size, false));

            return directories.Concat(files).ToList();
        }

        bool IsPathValid(string path) {
            Uri pathUri;
            Boolean isValidUri = Uri.TryCreate(path, UriKind.Absolute, out pathUri);
            return isValidUri && pathUri != null && pathUri.IsLoopback;
        }
    }
}

File: VB/EditorsDemo.Wpf/ModuleResources/CardWinExplorerDataSource.vb
Imports DevExpress.Xpf.Core
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Diagnostics
Imports System.Globalization
Imports System.IO
Imports System.Linq
Imports System.Windows.Input
Imports DevExpress.Utils.Helpers

Namespace EditorsDemo.ModuleResources
    Public Class CardWinExplorerDataSource
        Implements INotifyPropertyChanged

        Private Const Root As String = "Root"


        Private sizeType_Renamed As SizeIcon = SizeIcon.Medium

        Private path_Renamed As String

        Private searchText_Renamed As String
        Private ReadOnly UndoStack As New Stack(Of String)()
        Private ReadOnly NextStack As New Stack(Of String)()

        Private ReadOnly source_Renamed As New ObservableCollectionCore(Of CardWinExplorerFileSource)()

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Public Enum SizeIcon
            ExtraLarge
            Large
            Medium
        End Enum

        Public Property SizeType() As SizeIcon
            Get
                Return sizeType_Renamed
            End Get
            Set(ByVal value As SizeIcon)
                If value <> sizeType_Renamed Then
                    sizeType_Renamed = value
                    Resize()
                    If PropertyChangedEvent IsNot Nothing Then
                        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("Size"))
                        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("SizeType"))
                    End If
                End If
            End Set
        End Property

        Public ReadOnly Property Size() As Integer
            Get
                Select Case SizeType
                    Case SizeIcon.ExtraLarge
                        Return 256
                    Case SizeIcon.Large
                        Return 128
                    Case SizeIcon.Medium
                        Return 65
                    Case Else
                        Return 128
                End Select
            End Get
        End Property

        Public Property Path() As String
            Get
                Return path_Renamed
            End Get
            Set(ByVal value As String)
                Dim tempPath As String = value
                Dim fileExist As Boolean = File.Exists(value)
                If fileExist Then
                    tempPath = System.IO.Path.GetDirectoryName(tempPath)
                End If
                Dim directoryExist As Boolean = Directory.Exists(tempPath)

                If value <> Root AndAlso ((Not directoryExist) AndAlso (Not fileExist)) Then
                    OpenWeb(value)
                    RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("Path"))
                    Return
                End If
                If path_Renamed IsNot Nothing Then
                    UndoStack.Push(path_Renamed)
                End If
                path_Renamed = tempPath
                If fileExist Then
                    OpenFile(value)
                ElseIf directoryExist OrElse path_Renamed = Root Then
                    OpenFolder(tempPath)
                End If
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("Path"))
            End Set
        End Property
        Private Sub OpenFile(ByVal path As String)
            Try
                Process.Start(path)
            Catch
            End Try
        End Sub
        Private Sub OpenWeb(ByVal path As String)
            Try
                Process.Start("http://" & path & ".com")
            Catch
            End Try
        End Sub
        Public Property SearchText() As String
            Get
                Return searchText_Renamed
            End Get
            Set(ByVal value As String)
                searchText_Renamed = value
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("SearchText"))
            End Set
        End Property

        Public ReadOnly Property Source() As ObservableCollectionCore(Of CardWinExplorerFileSource)
            Get
                Return source_Renamed
            End Get
        End Property

        Public Property CurrentItem() As CardWinExplorerFileSource

        Public ReadOnly Property UndoCommand() As ICommand
            Get
                Return New DevExpress.Mvvm.DelegateCommand(AddressOf CallUndo, Function() UndoStack.Count > 0)
            End Get
        End Property

        Public ReadOnly Property NextCommand() As ICommand
            Get
                Return New DevExpress.Mvvm.DelegateCommand(AddressOf CallNext, Function() NextStack.Count > 0)
            End Get
        End Property

        Public ReadOnly Property OpenCommand() As ICommand
            Get
                Return New DevExpress.Mvvm.DelegateCommand(Sub() Open(CurrentItem), Function() CurrentItem IsNot Nothing)
            End Get
        End Property

        Public ReadOnly Property UpCommand() As ICommand
            Get
                Return New DevExpress.Mvvm.DelegateCommand(AddressOf CallUp, Function() Path <> Root)
            End Get
        End Property

        Public ReadOnly Property PropertyCommand() As ICommand
            Get
                Return New DevExpress.Mvvm.DelegateCommand(AddressOf CallExecute, Function() CurrentItem IsNot Nothing)
            End Get
        End Property

        Public Sub New()
            OpenRoot()
        End Sub

        Public Sub Open(ByVal element As CardWinExplorerFileSource)
            If element.Type = CardWinExplorerFileSource.TypeElement.File Then
                FileSystemHelper.ShellExecuteFileInfo(CurrentItem.FullPath, ShellExecuteInfoFileType.Open)
            Else
                Path = element.FullPath
            End If
        End Sub

        Private Sub OpenFolder(ByVal path As String, Optional ByVal clearNextStack As Boolean = True)
            Source.Clear()
            Source.BeginUpdate()
            Try
                Mouse.OverrideCursor = Cursors.Wait
                If path = Root Then
                    OpenRoot()
                Else

                    Dim sizeType_Renamed As SizeIcon = SizeType

                    Dim size_Renamed As Integer = Size
                    For Each item In FileSystemHelper.GetSubDirs(path)
                        Source.Add(New CardWinExplorerFileSource(System.IO.Path.Combine(path, item), CardWinExplorerFileSource.TypeElement.Folder, sizeType_Renamed, size_Renamed))
                    Next item
                    For Each item In Directory.EnumerateFiles(path)
                        Source.Add(New CardWinExplorerFileSource(item, CardWinExplorerFileSource.TypeElement.File, sizeType_Renamed, size_Renamed))
                    Next item
                End If
                If clearNextStack Then
                    NextStack.Clear()
                End If
            Catch ex As UnauthorizedAccessException
                System.Windows.MessageBox.Show(ex.Message, "Error", System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error)
                CallUndo()
            Finally
                Mouse.OverrideCursor = Nothing
                Source.EndUpdate()
            End Try
        End Sub

        Private Sub Resize()
            Try
                Mouse.OverrideCursor = Cursors.Wait
                CardWinExplorerFileSource.ClearCache()
                For Each item As CardWinExplorerFileSource In Source
                    item.Resize(SizeType, Size)
                Next item
            Finally
                Mouse.OverrideCursor = Nothing
            End Try
        End Sub

        Private Sub OpenRoot()
            Source.Clear()
            For Each drive In DriveInfo.GetDrives().Where(Function(x) x.DriveType = DriveType.Fixed)
                Source.Add(New CardWinExplorerFileSource(drive.RootDirectory.Name, CardWinExplorerFileSource.TypeElement.Drive, SizeType, Size))
            Next drive
            path_Renamed = Root
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("Path"))
        End Sub

        Private Sub CallUp()
            If Me.Path.Length <> 3 Then
                Path = System.IO.Directory.GetParent(Path).FullName
            Else
                Path = Root
            End If
        End Sub
        Private Sub CallExecute()
            FileSystemHelper.ShellExecuteFileInfo(CurrentItem.FullPath, ShellExecuteInfoFileType.Properties)
        End Sub

        Private Sub CallUndo()
            NextStack.Push(Path)
            Dim tmp As String = UndoStack.Pop()
            path_Renamed = tmp
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("Path"))
            OpenFolder(tmp, False)
        End Sub

        Private Sub CallNext()
            Dim tmp As String = NextStack.Pop()
            UndoStack.Push(path_Renamed)
            path_Renamed = tmp
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("Path"))
            OpenFolder(tmp, False)
        End Sub
        Public Function Find(ByVal path As String) As IEnumerable(Of CardWinExplorerFileSource)
            If path = Root Then
                Return DriveInfo.GetDrives().Where(Function(x) x.DriveType = DriveType.Fixed).Select(Function(drive) New CardWinExplorerFileSource(drive.RootDirectory.Name, CardWinExplorerFileSource.TypeElement.Drive, SizeType, Size, False))
            End If
            If String.IsNullOrEmpty(path) OrElse (Not IsPathValid(path)) Then
                Return Enumerable.Empty(Of CardWinExplorerFileSource)()
            End If
            If System.IO.Directory.Exists(path) Then
                path &= System.IO.Path.DirectorySeparatorChar
            End If
            Dim directory = System.IO.Path.GetDirectoryName(path)
            If String.IsNullOrEmpty(directory) Then
                directory = System.IO.Path.GetPathRoot(path)
            End If
            Dim filter As String = System.IO.Path.GetFileName(path)

            Dim directories = FileSystemHelper.GetSubDirs(directory).Where(Function(x) x.StartsWith(filter, True, CultureInfo.InvariantCulture)).Select(Function(x) New CardWinExplorerFileSource(System.IO.Path.Combine(directory, x), CardWinExplorerFileSource.TypeElement.Folder, SizeType, Size, False))
            Dim files = System.IO.Directory.EnumerateFiles(directory, filter & "*").Select(Function(x) New CardWinExplorerFileSource(x, CardWinExplorerFileSource.TypeElement.File, SizeType, Size, False))

            Return directories.Concat(files).ToList()
        End Function

        Private Function IsPathValid(ByVal path As String) As Boolean
            Dim pathUri As Uri = Nothing
            Dim isValidUri As Boolean = Uri.TryCreate(path, UriKind.Absolute, pathUri)
            Return isValidUri AndAlso pathUri IsNot Nothing AndAlso pathUri.IsLoopback
        End Function
    End Class
End Namespace
